File: problem-sets/leetcode-weekly-contest/20240316/3080-执行操作标记数组中的元素.py
from typing import List

class Solution:
    def unmarkedSumArray(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        n = len(nums)
        sorted_range = sorted(range(n), key=lambda x: nums[x])
        pt = 0
        vis = [0] * n
        tot = sum(nums)
        ans = []
        for i, k in queries:
            if vis[i] == 0:
                vis[i] = 1
                tot -= nums[i]
            for _ in range(k):
                while pt < n and vis[sorted_range[pt]]:
                    pt += 1
                # 指针移到最后，表示已经删完，需要退出循环，否则空跑循环是不满足复杂度要求的——不过力扣现在还没卡
                if pt  == n: break
                p = sorted_range[pt]
                vis[p] = 1
                tot -= nums[p]
            ans.append(tot)
        return ans

File: problem-sets/leetcode-weekly-contest/20240316/3080-执行操作标记数组中的元素.cpp
#include <numeric>
#include <queue>
#include <vector>

using std::vector;
using std::accumulate;
using std::pair;
using std::greater;
using std::priority_queue;

class Solution
{
public:
  vector<long long> unmarkedSumArray(vector<int> &nums, vector<vector<int>> &queries)
  {
    int n = nums.size();
    int m = queries.size();
    vector<long long> ans;
    long long sum = accumulate(nums.begin(), nums.end(), 0LL);
    vector<bool> marked(n, false);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    for (int i = 0; i < n; ++i) {
      pq.emplace(nums[i], i);
    }

    ans.reserve(m);
    for (auto &query : queries) {
      int index = query[0];
      int k     = query[1];

      if (!marked[index]) {
        marked[index]  = true;
        sum           -= nums[index];
      }

      while (k > 0 && !pq.empty()) {
        auto [num, idx] = pq.top();
        pq.pop();
        if (!marked[idx]) {
          marked[idx]  = true;
          sum         -= nums[idx];
          --k;
        }
      }
      ans.emplace_back(sum);
    }
    return ans;
  }
};